stacktrace

AlexGuteniev opened the pull request:
Resolves #1447

Thanks @cbezault for suggestion to use `<DbgEng.h>`, @fsb4000 for making sure it works on Windows 7, @ben-craig for pointing out that `<DbgEng.h>` does not look in a path next to the module by default.

Before going forward with this, here are some open issues:

### 1. API family, thread safety, COM apartments safety

There are three API families, neither of which seems fully good:

#### DbgHelp

The classic one. Every function has the following remark:

> All DbgHelp functions, such as this one, are single threaded. Therefore, calls from more than one thread to this function will likely result in unexpected behavior or memory corruption. To avoid this, you must synchronize all concurrent calls from more than one thread to this function.

This seem to exclude DbgHelp -- even though we can use locks internally, user may call the function bypassing the lock
On the good part, it does not have COM dependency.

#### DbgEng

`<DbgEng.h>` API is COM-like, but full COM. So still no COM dependency.

There's a problem with picking a `.pdb` next to image -- see below.

Also I've seen under a debugger that it uses `<DbgHelp>` underneath with locks. This means it is not thread safe too -- the client code may use DbgHelp directly.

There's a concert that the interfaces are documented as single-threaded and not usable via another thread. For now the problem is not seen, but if it is a problem, can be fixed by thread-local storage, also not caching these interfaces can be investigated.

#### DIA SDK

Uses COM. The usual usage of COM with `CoInitialize[Ex]` is problematic in STL - there are two incompatible COM modes, and if STL choices one, it will break programs which wants another (possibly in a subtle way).

There's information that `NoRegCoCreate` function can be used to avoid the need of `CoInitialize[Ex]`. I haven't tried yet. It is not a documented function, but it exists in DIA SDK headers (`DIA SDK\include\diacreate.h`) with a comment description.

DIA SDK is not a Windows component, but it is redistributable: https://docs.microsoft.com/en-us/visualstudio/releases/2022/redistribution#dia-sdk

#### 4. Roll own .pdb parsing

Tedious, but full control over thread and COM safety.
If we are to go this route, probably should make a separate open source library that can be reused elsewhere (Boost.Stacktrace, libc++ `std::stacktrace`, LLVM ASan stack trace)

#### 5. Use helper process

With this, DbgHelp will work fine. Though this won't be efficient

### 2. Looking up for `.pdb` next to image

As @ben-craig pointed out, seems like **DbgEng** does not work well when having `.pdb` in not exact path where it is expected. So having a `.pdb` next to EXE / DLL, but not in the same folder as where it was originally created makes symbol lookup failure.

Currently I have a workaround, `module_symbols_load_from_module_dir` function, but it is considered very harmful. The problems are:
 * Added paths can accumulate infinitely
 * Altering search path may cause loaded symbols to unload
Storing previous search dir and restoring it later will eliminate the former problem by making the later problem worse.

Apparently, with **DbgHelp** looking up next to image is done by default, and its [`SymSetOptions`](https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-symsetoptions) has `SYMOPT_IGNORE_IMAGEDIR` option to explicitly disable this. But **DbgEng** symbol does not list it among [its options](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/symbol-options).

Current resolution: use the workaround. Though I'd be more happy documenting the limitation and options to bypass it (like  `/PDBALTPATH:%_PDB%` as @Sylveon suggested).

### 3. Dll main safety

I cache pointers to `<DbgEng.h>` interfaces, removing them in `atexit`. for a DLL this is effectively in a `DllMain`. Though it is not safe to do most of thing in `DllMain`, I guess there are no other options. Not caching would be too inefficient, I think.
Need confirmation whether it is the way to go.

**Current resolution**: ignore the potential problem.

### Information format

I made the following:
 * `description`  is like `trace!f1+0x8a`. This includes module name, function name, offset within function (hex). No params, as they are hard to get. The paper does not ask for module name and offset within function, but I think these are essential. Note that `<DbgEng.h>` also provides module name by default, so it actually takes some effort to strip it.
 * `source_file` and `source_line` are obvious. Don't trim full path from source file if it is provided.
 * the whole `stacktrace_entry` is `file(line): description` to be compatible with Visual Studio usual output. Don't report offset within line ( `source_column`) as the paper does not ask for it, and it is not very essential.
 * finally stacktrace is entries separated by `\n` and prepended by `n>` to number entries starting from 0.

```
0> C:\Project\trace\st_try.cpp(5): P0881R7_stacktrace!innermost+0x35
1> C:\Project\trace\st_try.cpp(10): P0881R7_stacktrace!inner+0x17
2> C:\Project\trace\st_try.cpp(14): P0881R7_stacktrace!outer+0x17
3> C:\Project\trace\st_try.cpp(18): P0881R7_stacktrace!outermost+0x17
4> C:\Project\trace\st_try.cpp(22): P0881R7_stacktrace!main+0x17
5> d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(79): P0881R7_stacktrace!invoke_main+0x39
6> d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288): P0881R7_stacktrace!__scrt_common_main_seh+0x12e
7> d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(331): P0881R7_stacktrace!__scrt_common_main+0xe
8> d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp(17): P0881R7_stacktrace!mainCRTStartup+0xe
9> KERNEL32!BaseThreadInitThunk+0x14
10> ntdll!RtlUserThreadStart+0x21
```

@ben-craig points to Boost.Stacktrace has a different format, notably:
 * module/offset information is first, this makes output more aligned, as this information is always available
 * `#` symbol for entries umbering
 * `file.cpp:3` format of line numbering

This is similar to ASan output, like this:
```
    #0 0x7ff66557280f in innermost C:\Users\Aleksandr.Gutenev\source\repos\ConsoleApplication2\ConsoleApplication2\ConsoleApplication2.cpp:6
    #1 0x7ff665572844 in inner C:\Users\Aleksandr.Gutenev\source\repos\ConsoleApplication2\ConsoleApplication2\ConsoleApplication2.cpp:10
    #2 0x7ff665572864 in outer C:\Users\Aleksandr.Gutenev\source\repos\ConsoleApplication2\ConsoleApplication2\ConsoleApplication2.cpp:14
    #3 0x7ff665572884 in outermost C:\Users\Aleksandr.Gutenev\source\repos\ConsoleApplication2\ConsoleApplication2\ConsoleApplication2.cpp:19
    #4 0x7ff6655728a4 in main C:\Users\Aleksandr.Gutenev\source\repos\ConsoleApplication2\ConsoleApplication2\ConsoleApplication2.cpp:24
    #5 0x7ff6655736f8 in invoke_main d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #6 0x7ff6655735dd in __scrt_common_main_seh d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #7 0x7ff66557349d in __scrt_common_main d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:330
    #8 0x7ff66557378d in mainCRTStartup d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:16
    #9 0x7ffe4de97033 in BaseThreadInitThunk+0x13 (C:\WINDOWS\System32\KERNEL32.DLL+0x180017033)
    #10 0x7ffe4f8a2650 in RtlUserThreadStart+0x20 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x180052650)
```

The drawback is that IDE does not understand this format when trying to click on entry. There's DevCom-1660805 for that. If it is taken, then probably this format is better.

### Network symbols

Symbols can be loaded from network. Typically it is Microsoft Symbols Server, but could be any location where `_NT_SYMBOL_PATH` points to.

Should we load over network by default? Only when `_NT_SYMBOL_PATH` is like `SRV*c:\temp\WinSymbols*http://msdl.microsoft.com/download/symbols;` ? Or avoid network locations at all?

The primary concerns are:
 * Long waits
 * Security

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,305 @@
+        return _Frames.max_size();
+    }
+
+    [[nodiscard]] const_reference operator[](size_type _Sx) const {
```
strengthen?

AdamBucior commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,180 @@
+        wchar_t buffer[max_line_size];
+    };
+
+    static stacktrace_address_table_entry_t stacktrace_address_table[table_size];
```
Why is this global state needed?

AlexGuteniev replied:
Generally global state here needed for `OpenProcess`+`SymInitialize` / `SymCleanup`+`CloseHandle`.

This particular global state is a cache. The decision towards making a cache was guided by these facts:
 * Obtaining information should be lazy
 * We anyway have to lock, as APIs are not thread safe
 * We have a random-access array of stacktrace entries, which are returned by ref, meaning they should be plain pointers
 * My pas experience (although in Windows XP times) shows these APIs are slow
 * There's `source_file()` and `source_line()` which are likely to be called in sequence, so it is natural to cache at least one last entry
 * Making cache to be more than just one entry is almost free

AdamBucior replied:
I don't think it's useful to cache this information. I think that the main use for `stacktrace` is to help debugging. It is likely to be used only in case of exception. `stacktrace` doesn't have to be fast, it should be optimized for the case when it's never used (since most programs are meant not to exit with an exception). This cache is huge (if my calculations are correct it has 1024 entries each of which has an array of 2000 `wchar_t`s).

AlexGuteniev replied:
Removed the cache table, still cache last address.

AlexGuteniev replied:
Removed caching entirely.

AdamBucior commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,199 @@
+    auto line = __std_stacktrace_source_line(_Address);
+    auto desc = __std_stacktrace_description(_Address);
+
+    if (line != 0) {
+        result = std::format("{}({})", __std_stacktrace_source_file(_Address), line);
+    }
+
+    if (!desc.empty()) {
+        if (result.empty()) {
+            result = std::move(desc);
+        } else {
+            result = std::format("{}: {}", result, desc);
+        }
+    }
```
Wouldn't something like this be clearer and faster?
```suggestion
    if (!desc.empty()) {
        if (line == 0) {
            result = std::move(desc);
        } else {
            result = std::format("{}({}): {}", __std_stacktrace_source_file(_Address), line, desc);
        }
    }
```
It would even be possible to remove `result` altogether:
```suggestion
    if (!desc.empty()) {
        if (line == 0) {
            return desc;
        } else {
            return std::format("{}({}): {}", __std_stacktrace_source_file(_Address), line, desc);
        }
    }
```

AdamBucior commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,199 @@
+    return result;
+}
+
+[[nodiscard]] std::string __stdcall __std_stacktrace_to_string(const void* _Addresses, size_t _Size) {
```
Ditto, avoid locking for every entry

AlexGuteniev commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,199 @@
+#pragma comment(lib, "Dbghelp.lib")
+
+namespace {
+    static constexpr std::size_t max_line_size = 2000;
```
It is function name

AdamBucior replied:
2000 characters for function name sounds like an even bigger overkill than 2000 characters for file name

AlexGuteniev replied:
![spirit_crash](https://user-images.githubusercontent.com/3995422/150833395-f1841578-f62c-4d57-b919-c8ed9ee2a1c0.png)

AdamBucior replied:
Oh wow, this call stack looks absolutely terrifying

davidhunter22 replied:
Sorry if this question is dumb. I assume that the compiler throws away information about type aliases so you end up with hideous stuff like the above. Has there ever been proposals to keep this information so that errors and stack traces can be more like what the developer typed?

AlexGuteniev replied:
It was me doing post-mortem debugging having only a minidump. The crash was in some code that used `boost::spirit` back in 2009. Surprisingly I've managed to understand the issue after switching to WinDbg. 

The function names you see are normal function names for `boost::spirit`. Not sure if 2000 chars is enough.

AlexGuteniev replied:
@davidhunter22 , `std::basic_stacktrace<std::allocator<std::stacktrace_entry> >::current` is what I get currently. 

Even if we decide to parse out this to just `current`, the buffer is for full name, not trimmed.

AlexGuteniev replied:
And I don't think the implementation should do too smart processing, TBH

AlexGuteniev replied:
What enhancement I'm thinking of is adding parameters list. But I'm not sure whether it worth doing, it would add complexity.

AlexGuteniev replied:
As for the 2000 limitation, I'd like to avoid any limits at all (up to throwing bad alloc for function name of few gigabytes), but don't see a way to.

SuperWig replied:
That is quite something. I'd be interested in a word cloud of that 😅.

AlexGuteniev replied:
Ok, instead of arbitrary `static constexpr std::size_t max_line_size = 2000;` will use `MAX_SYM_NAME` defined as 2000 in `DbgHelp.h`

cbezault commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,232 @@
+#include <dbghelp.h>
+// clang-format on
+
+#pragma comment(lib, "Dbghelp.lib")
```
Why not use https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtlcapturestackbacktrace

AlexGuteniev replied:
I do use it for capturing, although under `CaptureStackBackTrace` name, the DbgHelp APIs are still needed for function names/line names/line numbers.

cbezault replied:
Yeah you're right

AlexGuteniev replied:
[DIA SDK](https://docs.microsoft.com/en-us/visualstudio/debugger/debug-interface-access/debug-interface-access-sdk?view=vs-2022) could be another option, but I'm afraid of using COM in STL impl.

AlexGuteniev commented:
After discussion with @cbezault , decided to switch from `DbgHelp.h` to `DbgEng.h`.
This should:
 * Address the concern about non-thread-safety
 * Not introduce COM apartments concerns (seems though `DbgEng.h` is COM, it does not need `CoInitialize`)
 * Eliminate satellite (should not be a problem to have distinct instances of `DbgEng.h` implementation)
   * Will use import library instead
   * `string` would be passed using a type-erasure wrapper to avoid having `string` in import library, but still avoid unnecessary allocations

StephanTLavavej commented:
Thanks for investigating this - I am **extremely** happy to hear that there will be no COM apartment and satellite DLL issues. With the thread safety concerns addressed, is the list of the questions in the PR description still up-to-date? We can start looking into them, but I wanted to make sure we'll be providing the answers you're looking for. :smile_cat:

AlexGuteniev commented:
The questions 3 and 5 are not related to which APIs used.

1 is sorted out by assuming dbgeng.h is safe
For 4 I'm now confident that my decision is correct 

2 is still valid, but needs to be restated with the new approach

crackedmind commented:
This is my two cents, dbghelp uses com interface inside from DIA SDK. DbgHelp uses backward compatibility.  And from docs

```
Although DbgHelp.dll ships with all versions of Windows, callers should consider using one of the more recent versions of this DLL as found in the [Debugging Tools For Windows](https://www.microsoft.com/?ref=go) package. For details on distribution of DbgHelp, see [DbgHelp Versions](https://docs.microsoft.com/en-us/windows/win32/debug/dbghelp-versions).
```
So, how you suggest to deal with it? distribute dbghelp with app or anything else?

AlexGuteniev commented:
@StephanTLavavej , I restated the questions taking into account `<DbgEng.h>` based implementation. I also removed question 4, as now it is clear that `RtlCaptureStackBackTrace` is the way to go.

@crackedmind , currently it is clear that neither DIA SDK not DbgHelp are good options, we will use DbgEng. I don't think any extra files should be deployed. Instead, should check if the currently minimally supported Windows version (I guess Windows 7) supports this usage. Note that `std::stacktrace` is not omnipotent debugging facility, rather it is minimal but portable utility, so we shouldn't need anything advanced that is only available with a DLL upgrade. (In particular, my past experience showed that distributing newer `DbgHelp.dll` helps to make `StackWalk` function better, but we use `RtlCaptureStackBackTrace` instead).

AdamBucior commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,274 @@
+        }
+
+        if (displacement != 0) {
+            constexpr size_t max_disp_num = std::size("+0x1111222233334444") - 1; // maximum possible line number
```
Could this be `"+0x1111222233334444"sv.size()` instead? Ditto below

AlexGuteniev replied:
Good point, but I'd want to stand away from `string_view` for this reason:
```C++
// This must be as small as possible, because its contents are
// injected into the msvcprt.lib and msvcprtd.lib import libraries.
// Do not include or define anything else here.
// In particular, basic_string must not be included here.
```

(See #2141 for more precise message)

AlexGuteniev commented:
I removed question 3. It was about inconsistency of the first function in the trace.
In debug, the first could have been the `current` itself:
```
C:\Project\STL\stl\inc\stacktrace(96): trace!std::basic_stacktrace<std::allocator<std::stacktrace_entry> >::current+0xa8
C:\Project\trace\trace.cpp(7): trace!f1+0x44
C:\Project\trace\trace.cpp(14): trace!f2+0x20
C:\Project\trace\trace.cpp(18): trace!f3+0x20
C:\Project\trace\trace.cpp(23): trace!main+0x20
d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(79): trace!invoke_main+0x39
d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288): trace!__scrt_common_main_seh+0x12e
d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(331): trace!__scrt_common_main+0xe
d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp(17): trace!mainCRTStartup+0xe
KERNEL32!BaseThreadInitThunk+0x14
ntdll!RtlUserThreadStart+0x21
```

After trying to write tests, turned out that it is not good, removed.

ddr533 commented:
Nice work!

AlexGuteniev commented:
@StephanTLavavej 

> We can start looking into them, but I wanted to make sure we'll be providing the answers you're looking for. 😸

Please do start. It takes a dive into source of DbgHelp / DbgEng / DIA SDK and/or contacting their developers, which I cannot do easily from my side.

ben-craig commented:
**DLLMain**
With the atexit behavior described, will that cause problems with unloading customer DLLs, or will it only cause problems when unloading an MSVC supplied DLL?  If it is only MSVC supplied DLLs, then maybe we can avoid deadlocks by self loading the MSVC DLL, preventing it from ever being unloaded.

While I think it _might_ be acceptable to prevent the unloading of MSVC supplied DLLS, I don't think it would be acceptable to prevent customer DLLs from unloading.

**Display format of a trace:**
For the format of the trace, I have a preference to put the module!function information before the file and line number.  This follows windbg and boost stacktrace precedents.  It also places the information that will always be there first in line.  This helps prevent horizontal scrolling back-and-forth when some source lines are present, but others are not.

This is just a customer viewpoint though, and not a maintainer viewpoint.  So weigh the suggestions accordingly.

AlexGuteniev commented:
> With the atexit behavior described, will that cause problems with unloading customer DLLs, or will it only cause problems when unloading an MSVC supplied DLL? 

Both. Up to the scenario where the STL is statically linked to customers DLL that should load/unload to arbitrary processes (like, a shell extension)

> For the format of the trace, I have a preference to put the module!function information before the file and line number. This follows windbg and boost stacktrace precedents. It also places the information that will always be there first in line. This helps prevent horizontal scrolling back-and-forth when some source lines are present, but others are not.

The format:
```
1>C:\Project\trace\st_try.cpp(5): P0881R7_stacktrace!innermost+0x35
1>C:\Project\trace\st_try.cpp(10): P0881R7_stacktrace!inner+0x17
1>C:\Project\trace\st_try.cpp(14): P0881R7_stacktrace!outer+0x17
1>C:\Project\trace\st_try.cpp(18): P0881R7_stacktrace!outermost+0x17
1>C:\Project\trace\st_try.cpp(22): P0881R7_stacktrace!main+0x17
1>d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(79): P0881R7_stacktrace!invoke_main+0x39
1>d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288): P0881R7_stacktrace!__scrt_common_main_seh+0x12e
1>d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(331): P0881R7_stacktrace!__scrt_common_main+0xe
1>d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp(17): P0881R7_stacktrace!mainCRTStartup+0xe
1>KERNEL32!BaseThreadInitThunk+0x14
1>ntdll!RtlUserThreadStart+0x21
```
Was originally selected as compatible with Visual Studio. If you run something as build step, and the IDE shows output, the navigation would be convenient (click on the line in output to jump to source). Boost.Test also uses `file(line): description` format for this reason.

Anyway what exactly new format do you propose? (I have function (`P0881R7_stacktrace!main`) and file name (`C:\Project\trace\st_try.cpp`) as strings and offset and line as unsigned integers to start with)

StephanTLavavej commented:
Comments from James McNellis:

> * If the STL takes a dependency on dbgeng, it should be a dynamic dependency (`LoadLibrary`), not a static dependency.  dbgeng is big and has a lot of dependencies.
> * Note that `dbgeng.dll` depends on `oleaut32.dll` which depends on `msvcp_win.dll`, so there may be additional dependency concerns there when Windows takes an LKG update with this feature.
> * I'd recommend inviting [internal mailing list] to review the usage of the dbgeng API.
> * The PR description suggests "Roll own `.pdb` parsing" as an alternative implementation option.  Please note that the PDB file format is an internal implementation detail of the Visual C++ toolset and is subject to change.  My team, which owns the PDB file format, would request that you definitely not consider doing this.  If you determine that this feature cannot be implemented with the existing, documented APIs, please let me know and we can discuss options.

("LKG update" refers to how Windows consumes the MSVC toolset.)

StephanTLavavej commented:
STL's tentative answers (before reviewing the code):

* API:
  + Looks like DbgEng is the best option; we can document our usage of it and the associated limitations. At least we'll "cooperate" with other users of DbgEng (although not anyone calling the underlying DbgHelp).
  + DIA SDK = COM = want to avoid. PDB parsing *might* be viable if absolutely necessary but we'd need to talk to James's team. Helper process is unprecedented in the STL and would seem to be extremely complicated.
* PDB path:
  + The current workaround sounds very problematic due to indefinite accumulation - the STL can be loaded by extremely long-running processes. Removing the workaround, documenting the path limitation, and suggesting options like `/PDBALTPATH` seems simpler.
* DLLMain safety:
  + I'm not sure that `atexit` is the right time for cleanup, as that's called during process unload (or UCRT unload which is typically the same). The STL is *usually* loaded for the lifetime of the process but this is not always the case. Good precedent here might be our C++20 `<chrono>` usage of `icu.dll`, where it looks like we just never unload ICU and the cached pointers remain valid for as long as the STL is alive.
  + Regarding caching of DbgEng interfaces - this may not be necessary at all, since `<stacktrace>` may not be performance-critical (most usage would seem to involve happening during exception or error handling, not 100k times a second). We still need to have the module loaded, but perhaps we don't need the usual system to cache function pointers? I don't have an inherent objection to caching function pointers though, as we have lots of precedent to follow.
* Information format:
  + Your current format looks fine to me, especially being clickable in the IDE. No change suggested (this is something that could be reworked later in response to user feedback if necessary).
* Network symbols:
  + Automatically loading over the network by default seems very risky (from a correctness and performance perspective, as the network can be extremely slow). @cbezault suggested that network loading could be an *option* as long as it's not the default (and we appropriately document the consequences of enabling it).

StephanTLavavej commented on tests/std/tests/P1502R1_standard_library_header_units/test.cpp:
```diff
@@ -68,6 +68,7 @@ import <span>;
 import <spanstream>;
 import <sstream>;
 import <stack>;
+import <stacktrace>;
```
Ideally, we'd additionally add a block below to exercise `<stacktrace>` machinery at runtime, below the `<stack>` test:

https://github.com/microsoft/STL/blob/4bb41db3d8e7605fc03656659752f3fad4133bc0/tests/std/tests/P1502R1_standard_library_header_units/test.cpp#L745-L759

However, due to `<stacktrace>`'s special requirements (needing a PDB etc.), it may be difficult to do so in this test. @michaelbprice suggests that a default-constructed `stacktrace_entry` might not have PDB requirements - that would be perfectly fine to test (as here we're just trying to exercise a little bit of each header). If that's not feasible, it would be reasonable to skip automated coverage of `<stacktrace>` here, but in that case, we should add a comment below along the lines of `// not testing <stacktrace> for now, due to its PDB requirements`.

AlexGuteniev replied:
It is feasible to construct even normal, not default-constructed `stacktrace_entry`.
Moreover, if we use `__declspec(dllexport)` functions, they will be seen in the trace, even if .pdb is missing, so we can check function names.
Sure file/line can be tested only with .pdb .
How far the test should go?

michaelbprice replied:
I don't think the test has to go too far. Even something as simple as showing that when imported as a module, you can even reference any of the exported names would be better than nothing. It looks like most of the existing tests go further than that and test behavior, but those other APIs are a lot easier create contrived examples for that have no external dependencies.

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,347 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _STACKSTRACE_
+#define _STACKSTRACE_
```
Typo: `_STACKSTRACE_` should be `_STACKTRACE_`. Occurs in the `#endif` comment too.

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,347 @@
+#pragma push_macro("new")
+#undef new
+
+using _Stacktrace_string_fill_callback = size_t (*)(char*, size_t, void* _Context);
+
+using _Stacktrace_string_fill = size_t (*)(size_t, void* _String, void* _Context, _Stacktrace_string_fill_callback);
+
+inline size_t _Stacktrace_string_fill_impl(
+    const size_t _Size, void* const _String, void* const _Context, const _Stacktrace_string_fill_callback _Callback) {
```
We generally define internal machinery within `std`, with the exception of separately compiled "flat C ABI" machinery which is global scope and `extern "C"`. Here, the typedefs `_Stacktrace_string_fill_callback` and `_Stacktrace_string_fill` are used by the `_EXTERN_C` functions below, so I believe they should appear within that block. However, I believe that the `_Stacktrace_string_fill_impl` could be moved below into `_STD_BEGIN` (and then it wouldn't need to qualify `_STD string`).

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,347 @@
+#pragma push_macro("new")
+#undef new
+
+using _Stacktrace_string_fill_callback = size_t (*)(char*, size_t, void* _Context);
+
+using _Stacktrace_string_fill = size_t (*)(size_t, void* _String, void* _Context, _Stacktrace_string_fill_callback);
```
To defend against the compiler options `/Gd /Gr /Gv /Gz` that change the default calling convention, function pointer typedefs in headers need to be marked with explicit calling conventions. I believe we now prefer `__stdcall` for this purpose. Any functions whose addresses are passed to parameters of such function pointer typedefs will also need to be marked with this calling convention (such as `_Stacktrace_string_fill_impl` below, I believe).

StephanTLavavej replied:
Suggestion: Right now, these typedefs provide names for some, but not all, parameters. It would seem to be more consistent to provide names for all parameters.

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,347 @@
+}
+
+_EXTERN_C
+unsigned short __stdcall __std_stacktrace_capture(unsigned long _FramesToSkip, unsigned long _FramesToCapture,
+    void** _BackTrace, unsigned long* _BackTraceHash) noexcept;
```
Naming style: We conventionally use `_Ugly_snakes` for function parameters, while `_UglyCamels` are used for template parameters.

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,347 @@
+unsigned short __stdcall __std_stacktrace_capture(unsigned long _FramesToSkip, unsigned long _FramesToCapture,
+    void** _BackTrace, unsigned long* _BackTraceHash) noexcept;
+
+// Some of these exports may throw (They would propagate bad_alloc potentially thrown from string::resize_and_overwrite)
```
These technically aren't exports as they're injected into the import lib; also I recommend adjusting the capitalization:

```suggestion
// Some of these functions may throw (they would propagate bad_alloc potentially thrown from string::resize_and_overwrite)
```

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,347 @@
+void __stdcall __std_stacktrace_source_file(
+    const void* _Address, void* _Str, _Stacktrace_string_fill) noexcept(false);
+
+[[nodiscard]] unsigned __stdcall __std_stacktrace_source_line(const void* _Address) noexcept;
+
+void __stdcall __std_stacktrace_to_string(const void* _Addresses, size_t _Size, void* _Str,
+    _Stacktrace_string_fill) noexcept(false);
+
+[[nodiscard]] void* __stdcall __std_stacktrace_get_debug_interface() noexcept;
```
These should be `_NODISCARD` as we're still using that macro (although we might unconditionally use the attribute in the future).

Additionally, we conventionally always say `unsigned int` instead of `unsigned`.

StephanTLavavej reviewed (commented):
Partial video review - still need to review `<stacktrace>` (at `_STD_BEGIN` and below), `stacktrace.cpp`, and the newly added `env.lst` and `test.cpp`. Reviewed all other files. Thanks for working on this complicated feature that users have been asking about! :heart_eyes_cat:

AlexGuteniev commented on stl/src/msvcp_stacktrace.def:
```diff
@@ -0,0 +1,10 @@
+; Copyright (c) Microsoft Corporation.
+; SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+EXPORTS
```
Yes, great catch. Will remove.

AlexGuteniev commented:
> PDB path:
> * The current workaround sounds very problematic due to indefinite accumulation - the STL can be loaded by extremely long-running processes. Removing the workaround, documenting the path limitation, and suggesting options like `/PDBALTPATH` seems simpler.

I really don't want this workaround.
But I hate having the issue of PDB path, and extra switches to fix it either.

I think the problem is only specific to DbgEng, but not DbgHelp or DIA. Thoughts:
 * **DbgHelp + satellite process** may be an option. Satellite process is new to STL, but if you can afford satellite DLLs, why not satellite processes.
 * DIA may be loaded without `CoCreateInstance` using its own factory function; in this case it falls into the same "fake COM" category as DbgEng.
 * If DbgEng can be fixed, there may be an option added to lift this limitation, then `/PDBALTPATH` will only be needed for older Windows only, which is somewhat better than always needing the workaround.

AlexGuteniev commented on stl/CMakeLists.txt:
```diff
@@ -457,7 +459,7 @@ function(add_stl_dlls D_SUFFIX THIS_CONFIG_DEFINITIONS THIS_CONFIG_COMPILE_OPTIO
 
     add_library(msvcp${D_SUFFIX} SHARED)
-    target_link_libraries(msvcp${D_SUFFIX} PRIVATE msvcp${D_SUFFIX}_eha_objects msvcp${D_SUFFIX}_objects msvcp${D_SUFFIX}_init_objects "${TOOLSET_LIB}/vcruntime${D_SUFFIX}.lib" "${TOOLSET_LIB}/msvcrt${D_SUFFIX}.lib" "ucrt${D_SUFFIX}.lib")
+    target_link_libraries(msvcp${D_SUFFIX} PRIVATE msvcp${D_SUFFIX}_eha_objects msvcp${D_SUFFIX}_objects msvcp${D_SUFFIX}_init_objects "${TOOLSET_LIB}/vcruntime${D_SUFFIX}.lib" "${TOOLSET_LIB}/msvcrt${D_SUFFIX}.lib" "ucrt${D_SUFFIX}.lib" "DbgEng.lib" "Shlwapi.lib")
```
Sure, will remove. But will wait until direction is clarified, to understand whether replace with dynamic DbgEng or something else.

monamimani commented:
I was interested to see own this feature was progressing and found really interesting the problems around it.
I read this comment from @StephanTLavavej  
` (most usage would seem to involve happening during exception or error handling, not 100k times a second)`

I am probably sure you though of it, but I actually think that there are 2 big use cases, the one you mentioned and where it will be most likely be called the most per seconds, is to keep track of where objects are created or resources acquired. At a later point in time ( at end of application for example), you print out a log of "live" objects/resources. That feature makes it really easy which one are "leaking" and where they got created/acquired. 
Those analytics are often enabled in release builds but not in Shipping builds. So even though they aren't that critical, reasonable performance would be best to allow that kind of usage.

FluXy commented:
Hi, don't know if it's useful or already mentioned somewhere else - this pdb reading lib was recently open sourced: https://github.com/MolecularMatters/raw_pdb

AlexGuteniev commented:
> Hi, don't know if it's useful or already mentioned somewhere else - this pdb reading lib was recently open sourced: https://github.com/MolecularMatters/raw_pdb

Thanks for the info, but I'm afraid we won't be looking into this library.
This part of the license:

https://github.com/MolecularMatters/raw_pdb/blob/66dccb0ba06952691c1fc0b1635f87f109099ee9/LICENSE#L12-L14

is incompatible with [Apache License v2.0 with LLVM Exceptions](https://llvm.org/releases/13.0.0/LICENSE.TXT)

StephanTLavavej commented on tests/std/tests/P0881R7_stacktrace/test.cpp:
```diff
@@ -0,0 +1,278 @@
+    assert(all.size() >= 4);
+
+#ifdef HAS_DEBUG_INFO
+    assert(filesystem::path(all.at(0).source_file()).filename() == "test.cpp"sv);
```
Should include `<string_view>`.

StephanTLavavej commented on tests/std/tests/P0881R7_stacktrace/test.cpp:
```diff
@@ -0,0 +1,278 @@
+    try {
+        (void) all.at(all.size());
+        assert(false); // should have thrown
+    } catch (out_of_range) {
```
Should include `<stdexcept>`. Also, `const out_of_range&` would avoid any concerns about copying exceptions.

StephanTLavavej commented on tests/std/tests/P0881R7_stacktrace/test.cpp:
```diff
@@ -0,0 +1,278 @@
+    assert(hash<stacktrace_entry>{}(all[0]) == hash<stacktrace_entry>{}(all_copy[0]));
+
+    assert(!all.empty());
+    assert(distance(all.begin(), all.end()) == static_cast<ptrdiff_t>(all.size()));
```
Should technically include `<iterator>` for `distance()` and `<cstddef>` for `ptrdiff_t`.

StephanTLavavej commented on tests/std/tests/P0881R7_stacktrace/test.cpp:
```diff
@@ -0,0 +1,278 @@
+}
+
+int main() {
+    std::thread t{test_impl};
+    test_impl();
+    t.join();
```
This test is already `using namespace std;`. Additionally, `jthread` would make the `join()` unnecessary.

AlexGuteniev commented on tests/std/tests/P1502R1_standard_library_header_units/test.cpp:
```diff
@@ -107,7 +108,8 @@ constexpr bool test_source_location() {
 }
 
-int main() {
+__declspec(dllexport) // for <stacktrace> test export main to have it named even without debug info
+    int main() {
```
Should also work, but preferred not to alter test matrices. Also debug info slows down linking noticeably.

StephanTLavavej replied:
Ah, ok - I think this is fine then. (This test takes long enough to build aready! :joy_cat:)

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,348 @@
+            void* _Context;
+        };
+
+        static_cast<string*>(_String)->resize_and_overwrite(_Size, _Stacktrace_string_fill_op{_Callback, _Context});
```
Yes, it is simpler with lambda, but I had code bloat concern. Do you think simplification worth using lambda?

AlexGuteniev replied:
Though I now agree that code bloat does not apply: even in /Od, where lambda is not inlined into call site, only one instantiation finally gets used. There's still some bloat of object files, but it may be not a big deal

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,348 @@
+private:
+    static constexpr size_t _Max_frames = 0xFFFF;
+
+    static constexpr bool _Noex_move = allocator_traits<_Al>::propagate_on_container_move_assignment::value
+                                    || allocator_traits<_Al>::is_always_equal::value;
```
clang-format was extremely confusing, so I avoided it this way

StephanTLavavej replied:
Makes sense, thanks!

AlexGuteniev commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,336 @@
+        srw_lock_guard& operator=(const srw_lock_guard&) = delete;
+
+    private:
+        SRWLOCK* const locked;
```
I want to simulate `std::muxet` so that it coild be rrplaced with `std::mutex` in vNext

AlexGuteniev commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,336 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// This must be as small as possible, because its contents are
+// injected into the msvcprt.lib and msvcprtd.lib import libraries.
+// Do not include or define anything else here.
+// In particular, basic_string must not be included here.
+
+#include <format>
```
:-(

AlexGuteniev commented on tests/std/tests/P0881R7_stacktrace/test.cpp:
```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
```
I can't imagine `<stacktrace>` not including `<string>`, but ok.

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,348 @@
+#pragma push_macro("new")
+#undef new
+
+using _Stacktrace_string_fill_callback = size_t(__stdcall*)(char* _Data, size_t _Size, void* _Context);
+
+using _Stacktrace_string_fill = size_t(__stdcall*)(
+    size_t _Size, void* _String, void* _Context, _Stacktrace_string_fill_callback _Callback);
+
+_EXTERN_C
```
extern "C" would imply `noexcept`, and without `__cpp_noexcept_function_type` cannot be overridden to `noexcept(false)`

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,348 @@
+#pragma push_macro("new")
+#undef new
+
+using _Stacktrace_string_fill_callback = size_t(__stdcall*)(char* _Data, size_t _Size, void* _Context);
+
+using _Stacktrace_string_fill = size_t(__stdcall*)(
+    size_t _Size, void* _String, void* _Context, _Stacktrace_string_fill_callback _Callback);
```
without `__cpp_noexcept_function_type` cannot be overridden to `noexcept(false)`

AlexGuteniev commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,336 @@
+        SRWLOCK* const locked;
+    };
+
+    IDebugClient* debug_client     = nullptr;
+    IDebugSymbols* debug_symbols   = nullptr;
+    IDebugSymbols3* debug_symbols3 = nullptr;
+    IDebugControl* debug_control   = nullptr;
+    bool attached                  = false;
+    bool initialize_attempted      = false;
+    SRWLOCK srw                    = SRWLOCK_INIT;
```
Defer till API questions resolution

StephanTLavavej commented:
I've pushed a merge with `main` to resolve a simple conflict in `stl/CMakeLists.txt` where #2654 added `"ole32.lib"`.

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,343 @@
+    source_file(_Address, _Str, 0, nullptr, _Fill);
+}
+
+unsigned int __stdcall __std_stacktrace_source_line(const void* const _Address) noexcept {
```
Should be `[[nodiscard]]` to match the declaration.

StephanTLavavej reviewed (commented):
Thanks! I think this should be ready for API review next. I'll validate and push changes for the remaining minor issues I noticed.

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,343 @@
+// Do not include or define anything else here.
+// In particular, basic_string must not be included here.
+
+#include <memory>
```
We'll also need `<yvals.h>` for `_STL_VERIFY`.

StephanTLavavej reviewed (commented):
Ready for API review. Current known issues:
* Static dependency on DbgEng should be dynamic.
* The static dependency also requires the Special Math and Atomic Wait satellite builds to be modified, which is undesirable.

StephanTLavavej commented:
The Debugging Experiences team reviewed this and had a few comments. My responses are sub-bullet-points:

* use of dbgeng is nice because it’s a lot less code than using dbghelp
* there are vague concerns around larger surface area
* What about existing applications that use the dbgeng – e.g. the application (which is itself a debugger) is not expecting to be debugging itself, or multiple processes in general.
  + (STL :cat:) Good point - I think we can resolve this through documentation.
* Are symbols resolved lazily? If so, that could lead to incorrect/wrong symbols when a dll gets unloaded after the backtrace was captured.
  + (STL :cat:) I believe the answer is that symbols are resolved when the user calls `std::to_string()` on their `std::stacktrace_entry` or `std::basic_stacktrace`, so it is indeed lazy/deferred. As this is inherent to the design of the C++ Standard’s API, I think we’ll simply need to document this. If a user is concerned about potential DLL unloads between the time of capture and resolution, they can immediately call `to_string()` and that will avoid any problems.

Thus I believe that the action items here are simply to convert the static dependency to dynamic, and to work with @TylerMSFT to ensure that the consequences of `<stacktrace>`'s dependency on DbgEng are properly documented.

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,344 @@
+    // __declspec(noinline) to make the same behavior for debug and release.
+    // We force the current function to be always noinline and add its frame to skipped.
+
+    _NODISCARD __declspec(noinline) static basic_stacktrace
+        current(const allocator_type& _Al = allocator_type()) noexcept {
+        basic_stacktrace _Result(_Internal_t{}, _Max_frames, _Al);
+        _Result._Frames.resize(__std_stacktrace_capture(1, static_cast<unsigned long>(_Max_frames),
+            reinterpret_cast<void**>(_Result._Frames.data()), &_Result._Hash));
+        return _Result;
+    }
+
+    _NODISCARD __declspec(noinline) static basic_stacktrace
+        current(size_type _Skip, const allocator_type& _Al = allocator_type()) noexcept {
+        basic_stacktrace _Result(_Internal_t{}, _Max_frames, _Al);
+        _Result._Frames.resize(__std_stacktrace_capture(static_cast<unsigned long>(_Skip + 1),
+            static_cast<unsigned long>(_Max_frames), reinterpret_cast<void**>(_Result._Frames.data()), &_Result._Hash));
+        return _Result;
+    }
+
+    _NODISCARD __declspec(noinline) static basic_stacktrace
+        current(size_type _Skip, size_type _Max_depth, const allocator_type& _Al = allocator_type()) noexcept {
+        basic_stacktrace _Result(_Internal_t{}, _Max_depth, _Al);
+        _Result._Frames.resize(__std_stacktrace_capture(static_cast<unsigned long>(_Skip + 1),
+            static_cast<unsigned long>(_Max_depth), reinterpret_cast<void**>(_Result._Frames.data()), &_Result._Hash));
+        return _Result;
+    }
```
The `current()` overloads are `noexcept`, but `basic_stacktrace` is constructing a non-empty `vector`, so allocation failure will terminate. [[stacktrace.basic.ctor]](https://eel.is/c++draft/stacktrace.basic.ctor) says that it should return "an empty `basic_stacktrace` object if the initialization of `frames_` failed".

AlexGuteniev commented:
> What about existing applications that use the dbgeng – e.g. the application (which is itself a debugger) is not expecting to be debugging itself, or multiple processes in general.

This is actually an argument towards helper process symbol resolver

StephanTLavavej commented:
@AlexGuteniev Now that 17.3 Preview 3 is branching for release, we'd like to get this ready to merge early in the 17.4 cycle (so that we can get early user feedback and respond if necessary). Are you planning to change the static dependency to dynamic in the near future, or should we find a maintainer to investigate that?

AlexGuteniev commented:
> Are you planning to change the static dependency to dynamic in the near future, or should we find a maintainer to investigate that?

I can do it over the next weekend.

Though I see the current usage of DbgEng problematioc, as it does not address locting `.pdb` next to `.exe` in case when both moved, foes not allow options for Internet symbols, and has issues with using in an app that is a debugger itself

AlexGuteniev commented:
Done dynamic import

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -85,49 +86,61 @@ namespace {
         }
 
+        if (dbgeng != nullptr) {
+            FreeLibrary(dbgeng);
```
[`FreeLibrary()`](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-freelibrary) returns a `BOOL` indicating success; ignoring it in `uninitialize()` seems reasonable, but we should probably say `(void)` to explicitly indicate that we're doing so (and to avoid code analysis warnings in the future).

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -85,49 +86,61 @@ namespace {
+            dbgeng = LoadLibraryW(L"dbgeng.dll");
+
+            if (dbgeng != nullptr) {
+                auto* debug_create = reinterpret_cast<decltype(&DebugCreate)>(GetProcAddress(dbgeng, "DebugCreate"));
```
We conventionally avoid `auto*` consistently now (it doesn't really provide any value over `auto`). This could also be top-level `const`.

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -85,49 +86,61 @@ namespace {
-                    0, GetCurrentProcessId(), DEBUG_ATTACH_NONINVASIVE | DEBUG_ATTACH_NONINVASIVE_NO_SUSPEND));
-                if (attached) {
-                    (void) debug_control->WaitForEvent(0, INFINITE);
+            dbgeng = LoadLibraryW(L"dbgeng.dll");
```
As `dbgeng.dll` lives in system32, I believe we need [`LoadLibraryExW()`](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexw) with  `LOAD_LIBRARY_SEARCH_SYSTEM32`:

https://github.com/microsoft/STL/blob/60decd0d829e68666b556780002c83605d748d80/stl/src/tzdb.cpp#L66

This avoids scary DLL planting issues.

strega-nil-ms reviewed (commented):
That's what I see right now; @stephantlavavej could you confirm that the names should be underscuglified?

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,360 @@
+        size_t _Size, void* _String, void* _Context, _Stacktrace_string_fill_callback _Callback);
+
+    template <class F>
+    size_t string_fill(const _Stacktrace_string_fill callback, const size_t size, void* const str, F f) {
```
Within `stl/src`, it isn't necessary to `_Uglify` the names of parameters, local variables, and other things that don't have external linkage. (Here, `string_fill()` is within an unnamed namespace.) We often use `_Ugly` names anyways - the rationale being that this is consistent with the headers where we spend the majority of our time, and that this makes it easier to move code between `stl/src` and `stl/inc` (rare, but it happens). However, we aren't 100% consistent and there are definitely many examples of non-`_Ugly` names, especially in older code. For example:

https://github.com/microsoft/STL/blob/60decd0d829e68666b556780002c83605d748d80/stl/src/xpoly.cpp#L10-L11

For non-external-linkage names, there is no risk of macroization or collision with user code, which is why this is OK. However, for external linkage (e.g. global scope or `std` scope functions, etc.) we need to use `_Ugly` names - even though we are isolated from user macros in `stl/src`, we don't want to collide at link time. (Which is why we're increasingly disciplined about using unnamed namespaces.)

The minimum bar we should meet is consistency within a single file - it appeared to me that `stacktrace.cpp` was consistently using non-`_Ugly` names except to match parameter names that also appeared in `stl/inc`, and that seemed fine.

I wouldn't object to using `_Ugly` names everywhere either, but I didn't think it was worth the effort.

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,360 @@
+        SRWLOCK* const locked;
+    };
+
+    IDebugClient* debug_client     = nullptr;
```
I was also somewhat concerned about how the SRWLOCK needs to protect all accesses to these variables. I agree that enforcing correctness by requiring access to go through the lock guard would be preferable, but I'm somewhat concerned about introducing a whole bunch of function calls into a facility that might be heavily used in non-optimized debug mode (specifically the `operator->()`). Also, introducing a `_Dbg_eng_data_t` and then capturing its address in a `_Lock_guard` would require us to pass that `_Lock_guard` between functions (by reference), since we want to take the lock exactly once per top-level function call, then access the guarded variables repeatedly, even when calling different helper functions.

I think there might be a different "correct by construction" approach:

* Move these variables (including the SRWLOCK) to be static data members of the existing `srw_lock_guard` marked `private`.
* Move the helper functions to be non-static member functions of the guard (like `try_initialize()`, `get_description()`, etc.)
* Then the top-level function calls would construct a guard and call those non-static member functions.

This way we wouldn't pay extra function calls, or have to pass around the address of a data-holder type - the variables are ultimately `static`, we have just one copy of them (hence the need for the SRWLOCK), and the guard object and non-static member functions are ensuring that we acquire/release the SRWLOCK before accessing the static data members.

I'm fine with the code as-is, though. It's also worth noting that because this is within an unnamed namespace within a single `stl/src` file, we can change it arbitrarily in the future without ABI concerns. (The behavior of the top-level `__std_meow` functions must remain unchanged, but they just need to all agree on taking the same lock, so there is no mix-and-match concern if we change the lock - they are within the same file.)

AlexGuteniev replied:
Remember that `SRWLOCK` will be `std::mutex` when `mutex` constructor would be `constexpr`, in this case it woild use anusual guard.

I think it is usual to just use locks responible, instead of trying to enforce them.

AlexGuteniev replied:
wrapped into structure

StephanTLavavej replied:
I've pushed changes to complete this improvement (making the member functions non-static, even though all of the state is static).

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,360 @@
+namespace {
+    // The below function pointer types must be in sync with <stacktrace>
+
+    using _Stacktrace_string_fill_callback = size_t(__stdcall*)(char* _Data, size_t _Size, void* _Context);
```
It would be possible, but it would have a small impact on user throughput (as `<stacktrace>` would then need to include a subheader). That seems excessive for two typedefs (whereas `xfilesystem_abi.h` contains a ton of structures and constants).

Some precedent - we've been willing to ship a small amount of duplication with comments that `inc` and `src` must match, e.g. in the `shared_mutex` implementation.

https://github.com/microsoft/STL/blob/60decd0d829e68666b556780002c83605d748d80/stl/inc/xthreads.h#L91-L99
https://github.com/microsoft/STL/blob/60decd0d829e68666b556780002c83605d748d80/stl/src/sharedmutex.cpp#L11-L20

My preference would be to leave this as-is, but if you feel strongly that we should avoid duplication going forward, I would not object to adding a subheader (the throughput impact is admittedly small, and a non-issue for Standard Library Modules which are the long-term future).

AlexGuteniev commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,360 @@
+    using _Stacktrace_string_fill_callback = size_t(__stdcall*)(char* _Data, size_t _Size, void* _Context);
+
+    using _Stacktrace_string_fill = size_t(__stdcall*)(
+        size_t _Size, void* _String, void* _Context, _Stacktrace_string_fill_callback _Callback);
```
besides `impl` is already used.

AlexGuteniev replied:
Added comment in the header.

strega-nil-ms commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,365 @@
+        _Result._Frames.resize(__std_stacktrace_capture(static_cast<unsigned long>(_Skip + 1),
+            static_cast<unsigned long>(_Max_frames), reinterpret_cast<void**>(_Result._Frames.data()), &_Result._Hash));
+        return _Result;
+        _CATCH_ALL
```
please split this into two calls; this is really difficult to understand without a named return value from `__std_stacktrace_capture`:

```cxx
        unsigned short actual_size = __std_stacktrace_capture(...);
        _Result._Frames.resize(actual_size);
```

(same comment above)

strega-nil-ms commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,365 @@
+};
+
+template <class _Alloc>
+class basic_stacktrace {
```
I think we need `static_assert`s here to make certain that `sizeof(stacktrace_entry) == sizeof(void*)`, and `alignof(stacktrace_entry) == alignof(void*)`, since we `reinterpret_cast<void**>(stacktrace_entry*)`. Additionally, I think a function call `to_voidptr_array` or smth that does that cast would be good, so we only have _one_ `reinterpret_cast` to look at.

strega-nil-ms reviewed (commented):
More comments; I haven't taken a look at the standardese just yet.

TylerMSFT commented:
Looks like this has been merged.

>> … work with @TylerMSFT<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FTylerMSFT&data=05%7C01%7CTyler.Whitney%40microsoft.com%7Cce38fc21e8bc4a57e85e08da2e24fca5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637873033644425837%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=v7HeTQyVcYNj4BudvLcFHAEDaLtH5lenvkvE8VnfNwQ%3D&reserved=0> to ensure that the consequences of <stacktrace>'s dependency on DbgEng are properly documented

Happy to update the docs. What are the details for the doc update?

-Tyler

From: Stephan T. Lavavej ***@***.***>
Sent: Wednesday, May 4, 2022 4:23 PM
To: microsoft/STL ***@***.***>
Cc: Tyler Whitney ***@***.***>; Mention ***@***.***>
Subject: Re: [microsoft/STL] stacktrace (PR #2502)


The Debugging Experiences team reviewed this and had a few comments. My responses are sub-bullet-points:

  *   use of dbgeng is nice because it’s a lot less code than using dbghelp
  *   there are vague concerns around larger surface area
  *   What about existing applications that use the dbgeng – e.g. the application (which is itself a debugger) is not expecting to be debugging itself, or multiple processes in general.
     *   (STL 🐱) Good point - I think we can resolve this through documentation.
  *   Are symbols resolved lazily? If so, that could lead to incorrect/wrong symbols when a dll gets unloaded after the backtrace was captured.
     *   (STL 🐱) I believe the answer is that symbols are resolved when the user calls std::to_string() on their std::stacktrace_entry or std::basic_stacktrace, so it is indeed lazy/deferred. As this is inherent to the design of the C++ Standard’s API, I think we’ll simply need to document this. If a user is concerned about potential DLL unloads between the time of capture and resolution, they can immediately call to_string() and that will avoid any problems.

Thus I believe that the action items here are simply to convert the static dependency to dynamic, and to work with @TylerMSFT<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FTylerMSFT&data=05%7C01%7CTyler.Whitney%40microsoft.com%7Cce38fc21e8bc4a57e85e08da2e24fca5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637873033644425837%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=v7HeTQyVcYNj4BudvLcFHAEDaLtH5lenvkvE8VnfNwQ%3D&reserved=0> to ensure that the consequences of <stacktrace>'s dependency on DbgEng are properly documented.

—
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2FSTL%2Fpull%2F2502%23issuecomment-1118021084&data=05%7C01%7CTyler.Whitney%40microsoft.com%7Cce38fc21e8bc4a57e85e08da2e24fca5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637873033644625792%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=k8qzPi%2F6KpCtcE2uFrl68aExEnk8FxLU6ghjeSYBeSY%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAC54FH7W2YQBOI3ECLCABSDVIMBEDANCNFSM5MTTQYWQ&data=05%7C01%7CTyler.Whitney%40microsoft.com%7Cce38fc21e8bc4a57e85e08da2e24fca5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637873033644625792%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=eso0UEWw9qn8yBT3JjPZ%2FLXHgz%2BPJwghhat%2B4Yk%2FjWg%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

StephanTLavavej commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,345 @@
+                }
+            }
+
+            if (std::atexit(lock_and_uninitialize) != 0) {
+                uninitialize();
+                return false;
+            }
```
We call `try_initialize()` repeatedly (once per `__std_stacktrace_meow()` call). This call to `std::atexit()` isn't guarded by `if (!initialize_attempted)`, so we'll repeatedly register `lock_and_uninitialize`. I believe that we'll eventually run out of slots in the atexit table, so this will fail.

Because `lock_and_uninitialize` is highly robust to partial initialization, I think that we can move the `std::atexit()` call within `if (!initialize_attempted)` very early, and then we wouldn't even need the `uninitialize()` call for `atexit` failure.

I will validate and push changes.

StephanTLavavej commented:
I pushed two commits:

* Call `atexit()` within `if (!initialize_attempted)`.
  + `lock_and_uninitialize()` is highly robust to partial initialization, so we can register it immediately.
  + Note that if `atexit()` fails, we haven't initialized anything, so we don't need to call `uninitialize()` to clean up. Instead, we can immediately return false.
  + Finally, note that if `atexit()` fails, we leave `initialize_attempted` set to true (`uninitialize()` would unset it), so we will never attempt initialization again.
* Mark `try_initialize()` as `[[nodiscard]]`; its return value is important.
  + Also use `[[nodiscard]]` instead of `_NODISCARD` for `dbg_eng_data`, to be consistent with other occurrences in this source file.

strega-nil-ms commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,344 @@
+            }
+
+            if (displacement != 0) {
+                constexpr size_t max_disp_num = sizeof("+0x1111222233334444") - 1; // maximum possible offset
```
```suggestion
                constexpr size_t max_disp_num = sizeof("+0x1122334455667788") - 1; // maximum possible offset
```

this obviously doesn't make a difference in code, but it makes obvious the byte boundaries.

strega-nil-ms replied:
Also, should this be `sizeof("+0x11223344")` on 32-bit platforms?

AlexGuteniev replied:
`sizeof(ULONG64) == 8`

In practice, PE format does not allow that much of code even for 64 bits, so there would be at most 8 hexits even in 64-bit, but I don't want to rely on such implicit limits, so used the type limit

strega-nil-ms replied:
ah, didn't realize they used `ULONG64` instead of `size_t`

strega-nil-ms reviewed (changes requested):
Mostly nits! Otherwise, this looks good :)

StephanTLavavej commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,384 @@
+        _TRY_BEGIN
+        basic_stacktrace _Result{_Internal_t{}, _Max_frames, _Al};
+        const unsigned short _Actual_size = __std_stacktrace_capture(
+            1, static_cast<unsigned long>(_Max_frames), _Result._To_voidptr_array(), &_Result._Hash);
```
We could file a followup issue to investigate whether 64K stack frames are realistic and whether we should detect and fail such cases. I believe we would be able to make any necessary changes in a compatible way, so I don't think this is critical to investigate for the initial merge.

AlexGuteniev replied:
https://docs.microsoft.com/windows/win32/debug/capturestackbacktrace

> ### FramesToCapture [in]
> 
> The number of frames to be captured.
> You can capture up to **MAXUSHORT** frames.

AlexGuteniev replied:
I think that capturing instead fewer frames is a better behavior than failing

strega-nil-ms replied:
We could capture more frames by just setting `skip` to `0xFFFF` (or maybe 0x10000`?) and capturing 0xFFFF more frames.

strega-nil-ms replied:
Resolved in favor of opening an issue.

AlexGuteniev commented on stl/src/stacktrace.cpp:
```diff
@@ -0,0 +1,344 @@
+_EXTERN_C
+[[nodiscard]] unsigned short __stdcall __std_stacktrace_capture(unsigned long _Frames_to_skip,
+    const unsigned long _Frames_to_capture, void** const _Back_trace, unsigned long* const _Back_trace_hash) noexcept {
+#ifdef _DEBUG
+    _Frames_to_skip += 1; // compensate for absence of tail call optimization here
+#endif
```
Don't we want better codegen for release builds?

strega-nil replied:
it's okay to have slightly worse codegen for one function, or alternatively to have slightly better codegen for one function

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,384 @@
+        explicit _Internal_t() noexcept = default;
+    };
+
+    basic_stacktrace(_Internal_t, size_type _Max_depth, const allocator_type& _Al) : _Frames(_Max_depth, _Al) {}
```
Can be considered as a rudiment. Or an optimization: we don't re-initialize vector, just initialize it right from the beginning.

AlexGuteniev replied:
Getting rid of this constructor and separating initial `resize` can help rewriting this with single return, which might assist RVO.

strega-nil-ms replied:
Resolving; I don't mind either way.

AlexGuteniev commented on stl/inc/stacktrace:
```diff
@@ -0,0 +1,384 @@
+_NODISCARD unsigned int __stdcall __std_stacktrace_source_line(const void* _Address) noexcept;
+
+void __stdcall __std_stacktrace_to_string(
+    const void* _Addresses, size_t _Size, void* _Str, _Stacktrace_string_fill _Fill) noexcept(false);
```
cannot be shared with `_To_voidptr_array()` due to const correctness.
`_To_voidptr_array()` is not const, and the return type does not have middle const.

strega-nil-ms reviewed (approved):
Approved! If you want, you can also change to using `resize` if you like.

strega-nil-ms commented:
@StephanTLavavej - I have not fully researched `DbgEng.h`; I feel comfortable with `<stacktrace>`, but don't necessarily fully know that `stacktrace.cpp` is completely correct. If you feel comfortable with `stacktrace.cpp` as it's written, please merge; else, I want someone to actually go fully through `stacktrace.cpp`.

StephanTLavavej commented:
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

StephanTLavavej commented:
I have pushed an additional commit to fix the internal test harness.

StephanTLavavej commented:
Thanks for implementing this **major** C++23 feature!

:heart_eyes_cat: :tada: :sparkles:
===

TylerMSFT commented:
@AlexGuteniev , @StephanTLavavej , it looked like there was a doc ask here. What are the details?

AlexGuteniev commented:
#2779 documnent limits in my take on limits. Still want that to be reviewed by maintainers first
